apps/openmw/mwmechanics/spellcasting.cpp

```diff
@@ -305,6 +270,77 @@ namespace MWMechanics
+    void CastSpell::getProjectileInfoAndLaunch (const ESM::EffectList& effects)
+    {
+        // All projectiles should use the same speed. From observations in the
+        // original engine, this seems to be the average of the constituent effects.
+        // First we get this average speed.
+        float speed = 0;
+        int count = 0;
+        for (std::vector<ESM::ENAMstruct>::const_iterator iter (effects.mList.begin());
+            iter!=effects.mList.end(); ++iter)
+        {
+            const ESM::MagicEffect *magicEffect = MWBase::Environment::get().getWorld()->getStore().get<ESM::MagicEffect>().find (
+                iter->mEffectID);
+            speed += magicEffect->mData.mSpeed;
+            count++;
+        }
+
+        if (count != 0)
+            speed /= count;
+
+        std::string model;
+        std::string sound;
+        
+        osg::Vec3f fallbackDirection (0,1,0);
+
+        bool isFirstProjectile = true;
+
+        for (std::vector<ESM::ENAMstruct>::const_iterator iter (effects.mList.begin());
+            iter!=effects.mList.end(); ++iter)
+        {
+            if (iter->mRange != ESM::RT_Target)
+                continue;
+
+            const ESM::MagicEffect *magicEffect = MWBase::Environment::get().getWorld()->getStore().get<ESM::MagicEffect>().find (
+                iter->mEffectID);
+
+            model = magicEffect->mBolt;
+            if (model.empty())
+                model = "VFX_DefaultBolt";
+
+            static const std::string schools[] = {
+                "alteration", "conjuration", "destruction", "illusion", "mysticism", "restoration"
+            };
+            if (!magicEffect->mBoltSound.empty())
+                sound = magicEffect->mBoltSound;
+            else
+                sound = schools[magicEffect->mData.mSchool] + " bolt";
+
+            // Fall back to a "caster to target" direction if we have no other means of determining it
+            // (e.g. when cast by a non-actor)
+            if (!mTarget.isEmpty())
+                fallbackDirection =
+                   osg::Vec3f(mTarget.getRefData().getPosition().asVec3())-
+                   osg::Vec3f(mCaster.getRefData().getPosition().asVec3());
+            
+            // Only send the effects data with the first projectile, so we don't have the impact sounds
+            // playing multiple times.
+            if (isFirstProjectile)
+                MWBase::Environment::get().getWorld()->launchMagicBolt(model, sound, mId, speed,
+                                                       false, effects, mCaster, mSourceName, fallbackDirection);
+            else
+            {
+                ESM::EffectList empty;
+                MWBase::Environment::get().getWorld()->launchMagicBolt(model, sound, mId, speed,
+                                                       false, empty, mCaster, mSourceName, fallbackDirection);
```

Comment:
Launching a projectile with no effects seems like a hack. I think it'd make more sense to allow a projectile to have more than one sound / model (just like it can have multiple effects). I.e. you would supply a list of sounds & list of models to launchMagicBolt.